File: src/lib/projects.ts
import { invoke } from "@tauri-apps/api/core";
import type { Project, ProjectSnapshot, Session } from "../types";
import { getBackendMode } from "./backend";

const BROWSER_STORAGE_KEY = "butler.projects.v1";
const DEFAULT_PROJECT: Project = {
  id: 0,
  name: "Unassigned",
  remotePath: null,
  sortOrder: 0,
  isDefault: true,
};

let browserSnapshot = loadBrowserSnapshot();

export async function getProjectSnapshot(): Promise<ProjectSnapshot> {
  if (getBackendMode() === "browser-preview") {
    return cloneSnapshot(browserSnapshot);
  }

  return invoke<ProjectSnapshot>("project_snapshot");
}

export async function createProject(
  name: string,
  remotePath: string | null,
): Promise<ProjectSnapshot> {
  if (getBackendMode() === "browser-preview") {
    const cleanName = validateName(name);
    ensureUniqueName(browserSnapshot.projects, cleanName);
    const id = nextProjectId(browserSnapshot.projects);
    const sortOrder =
      Math.max(
        0,
        ...browserSnapshot.projects.map((project) => project.sortOrder),
      ) + 1;
    browserSnapshot.projects.push({
      id,
      name: cleanName,
      remotePath: normalizeRemotePath(remotePath),
      sortOrder,
      isDefault: false,
    });
    persistBrowserSnapshot();
    return cloneSnapshot(browserSnapshot);
  }

  return invoke<ProjectSnapshot>("create_project", {
    name,
    remotePath,
  });
}

export async function updateProject(
  projectId: number,
  name: string,
  remotePath: string | null,
): Promise<ProjectSnapshot> {
  if (getBackendMode() === "browser-preview") {
    const project = browserSnapshot.projects.find(
      (candidate) => candidate.id === projectId,
    );
    if (!project) {
      throw new Error(`Project ${projectId} no longer exists.`);
    }
    const cleanName = validateName(name);
    ensureUniqueName(browserSnapshot.projects, cleanName, projectId);
    project.name = cleanName;
    project.remotePath = normalizeRemotePath(remotePath);
    persistBrowserSnapshot();
    return cloneSnapshot(browserSnapshot);
  }

  return invoke<ProjectSnapshot>("update_project", {
    projectId,
    name,
    remotePath,
  });
}

export async function deleteProject(projectId: number): Promise<ProjectSnapshot> {
  if (getBackendMode() === "browser-preview") {
    if (projectId === DEFAULT_PROJECT.id) {
      throw new Error("The default project cannot be deleted.");
    }
    if (!browserSnapshot.projects.some((project) => project.id === projectId)) {
      throw new Error(`Project ${projectId} no longer exists.`);
    }
    browserSnapshot.projects = browserSnapshot.projects.filter(
      (project) => project.id !== projectId,
    );
    browserSnapshot.assignments = Object.fromEntries(
      Object.entries(browserSnapshot.assignments).filter(
        ([, assignedProjectId]) => assignedProjectId !== projectId,
      ),
    );
    persistBrowserSnapshot();
    return cloneSnapshot(browserSnapshot);
  }

  return invoke<ProjectSnapshot>("delete_project", { projectId });
}

export async function assignSessionProject(
  sessionId: string,
  projectId: number,
): Promise<ProjectSnapshot> {
  if (getBackendMode() === "browser-preview") {
    if (!browserSnapshot.projects.some((project) => project.id === projectId)) {
      throw new Error(`Project ${projectId} no longer exists.`);
    }
    if (projectId === DEFAULT_PROJECT.id) {
      delete browserSnapshot.assignments[sessionId];
    } else {
      browserSnapshot.assignments[sessionId] = projectId;
    }
    persistBrowserSnapshot();
    return cloneSnapshot(browserSnapshot);
  }

  return invoke<ProjectSnapshot>("assign_session_project", {
    sessionId,
    projectId,
  });
}

export function applyProjectSnapshot(
  sessions: readonly Session[],
  snapshot: ProjectSnapshot,
): Session[] {
  const projects = new Map(
    snapshot.projects.map((project) => [project.id, project]),
  );
  const defaultProject =
    snapshot.projects.find((project) => project.isDefault) ??
    snapshot.projects[0] ??
    DEFAULT_PROJECT;

  return sessions.map((session) => {
    const assignedProjectId =
      snapshot.assignments[session.oodSessionId] ?? defaultProject.id;
    const project = projects.get(assignedProjectId) ?? defaultProject;
    return {
      ...session,
      projectId: project.id,
      projectName: project.name,
      remotePath: project.remotePath,
    };
  });
}

function loadBrowserSnapshot(): ProjectSnapshot {
  try {
    const stored = window.localStorage.getItem(BROWSER_STORAGE_KEY);
    if (!stored) {
      return { projects: [{ ...DEFAULT_PROJECT }], assignments: {} };
    }
    const parsed = JSON.parse(stored) as Partial<ProjectSnapshot>;
    const projects = Array.isArray(parsed.projects)
      ? parsed.projects.filter(isProject).map((project) => ({ ...project }))
      : [];
    if (
      !projects.some(
        (project) =>
          project.isDefault || project.id === DEFAULT_PROJECT.id,
      )
    ) {
      projects.unshift({ ...DEFAULT_PROJECT });
    }
    for (const project of projects) {
      project.isDefault = project.id === DEFAULT_PROJECT.id;
    }
    projects.sort(
      (left, right) => left.sortOrder - right.sortOrder || left.id - right.id,
    );
    const validProjectIds = new Set(projects.map((project) => project.id));
    const assignments = Object.fromEntries(
      Object.entries(parsed.assignments ?? {}).filter(
        ([sessionId, projectId]) =>
          sessionId.length > 0 &&
          typeof projectId === "number" &&
          projectId !== DEFAULT_PROJECT.id &&
          validProjectIds.has(projectId),
      ),
    );
    return { projects, assignments };
  } catch {
    return { projects: [{ ...DEFAULT_PROJECT }], assignments: {} };
  }
}

function persistBrowserSnapshot(): void {
  browserSnapshot.projects.sort(
    (left, right) => left.sortOrder - right.sortOrder || left.id - right.id,
  );
  try {
    window.localStorage.setItem(
      BROWSER_STORAGE_KEY,
      JSON.stringify(browserSnapshot),
    );
  } catch {
    // Browser preview persistence is best effort; Tauri uses the Rust metadata store.
  }
}

function cloneSnapshot(snapshot: ProjectSnapshot): ProjectSnapshot {
  return {
    projects: snapshot.projects.map((project) => ({ ...project })),
    assignments: { ...snapshot.assignments },
  };
}

function nextProjectId(projects: readonly Project[]): number {
  return Math.max(0, ...projects.map((project) => project.id)) + 1;
}

function validateName(name: string): string {
  const cleanName = name.trim();
  if (!cleanName) {
    throw new Error("Project name cannot be empty.");
  }
  if ([...cleanName].length > 80) {
    throw new Error("Project names must be 80 characters or fewer.");
  }
  return cleanName;
}

function ensureUniqueName(
  projects: readonly Project[],
  name: string,
  exceptProjectId?: number,
): void {
  if (
    projects.some(
      (project) =>
        project.id !== exceptProjectId &&
        project.name.toLocaleLowerCase() === name.toLocaleLowerCase(),
    )
  ) {
    throw new Error(`A project named “${name}” already exists.`);
  }
}

function normalizeRemotePath(remotePath: string | null): string | null {
  const value = remotePath?.trim() ?? "";
  if (!value) {
    return null;
  }
  if ([...value].length > 1_024) {
    throw new Error("Remote folders must be 1,024 characters or fewer.");
  }
  return value;
}

function isProject(value: unknown): value is Project {
  if (!value || typeof value !== "object") {
    return false;
  }
  const project = value as Partial<Project>;
  return (
    typeof project.id === "number" &&
    typeof project.name === "string" &&
    (project.remotePath === null || typeof project.remotePath === "string") &&
    typeof project.sortOrder === "number" &&
    typeof project.isDefault === "boolean"
  );
}

File: src-tauri/src/projects.rs
use std::{
    collections::{HashMap, HashSet},
    fs,
    path::{Path, PathBuf},
    sync::{Arc, Mutex},
};

use serde::{Deserialize, Serialize};

use crate::models::{Project, ProjectSnapshot};

const DEFAULT_PROJECT_ID: i64 = 0;
const PROJECTS_FILE_NAME: &str = "projects.json";

#[derive(Clone)]
pub struct ProjectState {
    inner: Arc<ProjectStateInner>,
}

struct ProjectStateInner {
    path: PathBuf,
    data: Mutex<ProjectData>,
}

#[derive(Clone, Serialize, Deserialize)]
struct ProjectData {
    #[serde(default = "default_next_project_id")]
    next_project_id: i64,
    #[serde(default)]
    projects: Vec<Project>,
    #[serde(default)]
    assignments: HashMap<String, i64>,
}

impl Default for ProjectData {
    fn default() -> Self {
        Self {
            next_project_id: 1,
            projects: vec![default_project()],
            assignments: HashMap::new(),
        }
    }
}

impl ProjectState {
    pub fn load(app_config_dir: &Path) -> Result<Self, String> {
        fs::create_dir_all(app_config_dir).map_err(|error| {
            format!(
                "Could not create Butler's application data directory {}: {error}",
                app_config_dir.display()
            )
        })?;

        let path = app_config_dir.join(PROJECTS_FILE_NAME);
        let mut data = match fs::read(&path) {
            Ok(contents) => serde_json::from_slice(&contents).map_err(|error| {
                format!(
                    "Could not parse Butler project metadata at {}: {error}",
                    path.display()
                )
            })?,
            Err(error) if error.kind() == std::io::ErrorKind::NotFound => ProjectData::default(),
            Err(error) => {
                return Err(format!(
                    "Could not read Butler project metadata at {}: {error}",
                    path.display()
                ))
            }
        };

        repair_loaded_data(&mut data);
        save(&path, &data)?;

        Ok(Self {
            inner: Arc::new(ProjectStateInner {
                path,
                data: Mutex::new(data),
            }),
        })
    }

    pub fn snapshot(&self) -> Result<ProjectSnapshot, String> {
        let data = self.data()?;
        Ok(snapshot(&data))
    }

    pub fn create_project(
        &self,
        name: String,
        remote_path: Option<String>,
    ) -> Result<ProjectSnapshot, String> {
        let name = clean_name(&name)?;
        let remote_path = clean_path(remote_path)?;
        let mut data = self.data()?;
        ensure_unique_name(&data.projects, &name, None)?;

        let id = data.next_project_id.max(1);
        data.next_project_id = id.saturating_add(1);
        let sort_order = data
            .projects
            .iter()
            .map(|project| project.sort_order)
            .max()
            .unwrap_or(0)
            .saturating_add(1);
        data.projects.push(Project {
            id,
            name,
            remote_path,
            sort_order,
            is_default: false,
        });
        save(&self.inner.path, &data)?;
        Ok(snapshot(&data))
    }

    pub fn update_project(
        &self,
        project_id: i64,
        name: String,
        remote_path: Option<String>,
    ) -> Result<ProjectSnapshot, String> {
        let name = clean_name(&name)?;
        let remote_path = clean_path(remote_path)?;
        let mut data = self.data()?;
        ensure_unique_name(&data.projects, &name, Some(project_id))?;

        let project = data
            .projects
            .iter_mut()
            .find(|project| project.id == project_id)
            .ok_or_else(|| format!("Project {project_id} no longer exists"))?;
        project.name = name;
        project.remote_path = remote_path;
        save(&self.inner.path, &data)?;
        Ok(snapshot(&data))
    }

    pub fn delete_project(&self, project_id: i64) -> Result<ProjectSnapshot, String> {
        if project_id == DEFAULT_PROJECT_ID {
            return Err("The default project cannot be deleted".into());
        }

        let mut data = self.data()?;
        let project_count = data.projects.len();
        data.projects.retain(|project| project.id != project_id);
        if data.projects.len() == project_count {
            return Err(format!("Project {project_id} no longer exists"));
        }
        data.assignments
            .retain(|_, assigned_project_id| *assigned_project_id != project_id);
        save(&self.inner.path, &data)?;
        Ok(snapshot(&data))
    }

    pub fn assign_session(
        &self,
        session_id: String,
        project_id: i64,
    ) -> Result<ProjectSnapshot, String> {
        let session_id = session_id.trim();
        if session_id.is_empty() {
            return Err("The session ID is invalid".into());
        }

        let mut data = self.data()?;
        if !data.projects.iter().any(|project| project.id == project_id) {
            return Err(format!("Project {project_id} no longer exists"));
        }
        if project_id == DEFAULT_PROJECT_ID {
            data.assignments.remove(session_id);
        } else {
            data.assignments.insert(session_id.to_owned(), project_id);
        }
        save(&self.inner.path, &data)?;
        Ok(snapshot(&data))
    }

    fn data(&self) -> Result<std::sync::MutexGuard<'_, ProjectData>, String> {
        self.inner
            .data
            .lock()
            .map_err(|_| "The project metadata store is unavailable".to_string())
    }
}

fn default_next_project_id() -> i64 {
    1
}

fn default_project() -> Project {
    Project {
        id: DEFAULT_PROJECT_ID,
        name: "Unassigned".into(),
        remote_path: None,
        sort_order: 0,
        is_default: true,
    }
}

fn repair_loaded_data(data: &mut ProjectData) {
    if !data
        .projects
        .iter()
        .any(|project| project.id == DEFAULT_PROJECT_ID)
    {
        data.projects.push(default_project());
    }

    for project in &mut data.projects {
        project.is_default = project.id == DEFAULT_PROJECT_ID;
    }
    data.projects.sort_by_key(|project| (project.sort_order, project.id));

    let project_ids = data
        .projects
        .iter()
        .map(|project| project.id)
        .collect::<HashSet<_>>();
    data.assignments.retain(|session_id, project_id| {
        !session_id.trim().is_empty()
            && *project_id != DEFAULT_PROJECT_ID
            && project_ids.contains(project_id)
    });

    let next_id = data
        .projects
        .iter()
        .map(|project| project.id)
        .max()
        .unwrap_or(0)
        .saturating_add(1);
    data.next_project_id = data.next_project_id.max(next_id).max(1);
}

fn clean_name(name: &str) -> Result<String, String> {
    let name = name.trim();
    if name.is_empty() {
        return Err("Project name cannot be empty".into());
    }
    if name.chars().count() > 80 {
        return Err("Project names must be 80 characters or fewer".into());
    }
    Ok(name.to_owned())
}

fn clean_path(path: Option<String>) -> Result<Option<String>, String> {
    let Some(path) = path else {
        return Ok(None);
    };
    let path = path.trim();
    if path.is_empty() {
        return Ok(None);
    }
    if path.chars().count() > 1_024 {
        return Err("Remote folders must be 1,024 characters or fewer".into());
    }
    Ok(Some(path.to_owned()))
}

fn ensure_unique_name(
    projects: &[Project],
    name: &str,
    except_project_id: Option<i64>,
) -> Result<(), String> {
    if projects.iter().any(|project| {
        Some(project.id) != except_project_id && project.name.eq_ignore_ascii_case(name)
    }) {
        return Err(format!("A project named “{name}” already exists"));
    }
    Ok(())
}

fn snapshot(data: &ProjectData) -> ProjectSnapshot {
    ProjectSnapshot {
        projects: data.projects.clone(),
        assignments: data.assignments.clone(),
    }
}

fn save(path: &Path, data: &ProjectData) -> Result<(), String> {
    let contents = serde_json::to_vec_pretty(data)
        .map_err(|error| format!("Could not serialize Butler project metadata: {error}"))?;
    fs::write(path, contents).map_err(|error| {
        format!(
            "Could not save Butler project metadata at {}: {error}",
            path.display()
        )
    })
}

#[cfg(test)]
mod tests {
    use std::time::{SystemTime, UNIX_EPOCH};

    use super::*;

    #[test]
    fn projects_and_assignments_survive_reload() {
        let unique = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap()
            .as_nanos();
        let directory = std::env::temp_dir().join(format!(
            "butler-projects-{}-{unique}",
            std::process::id()
        ));

        let state = ProjectState::load(&directory).unwrap();
        let snapshot = state
            .create_project("Research".into(), Some("~/research".into()))
            .unwrap();
        let project_id = snapshot
            .projects
            .iter()
            .find(|project| project.name == "Research")
            .unwrap()
            .id;
        state
            .assign_session("session-a".into(), project_id)
            .unwrap();

        let reloaded = ProjectState::load(&directory).unwrap().snapshot().unwrap();
        assert_eq!(reloaded.assignments.get("session-a"), Some(&project_id));
        assert!(reloaded.projects.iter().any(|project| project.is_default));
        assert!(reloaded
            .projects
            .iter()
            .any(|project| project.name == "Research"));

        let _ = fs::remove_dir_all(directory);
    }
}
